jax/_src/scipy/signal.py

```diff
@@ -146,3 +151,336 @@ def detrend(data, axis=-1, type='linear', bp=0, overwrite_data=None):
+    noverlap = int(noverlap)
+  if noverlap >= nperseg:
+    raise ValueError('noverlap must be less than nperseg.')
+  nstep = nperseg - noverlap
+
+  # Apply paddings
+  if boundary is not None:
+    ext_func = boundary_funcs[boundary]
+    x = ext_func(x, nperseg // 2, axis=-1)
+    if y is not None:
+      y = ext_func(y, nperseg // 2, axis=-1)
+
+  if padded:
+    # Pad to integer number of windowed segments
+    # I.e make x.shape[-1] = nperseg + (nseg-1)*nstep, with integer nseg
+    nadd = (-(x.shape[-1]-nperseg) % nstep) % nperseg
+    zeros_shape = list(x.shape[:-1]) + [nadd]
+    x = jnp.concatenate((x, jnp.zeros(zeros_shape)), axis=-1)
+    if y is not None:
+      zeros_shape = list(y.shape[:-1]) + [nadd]
+      y = jnp.concatenate((y, jnp.zeros(zeros_shape)), axis=-1)
+
+  # Handle detrending and window functions
+  if not detrend_type:
+    def detrend_func(d):
+      return d
+  elif not hasattr(detrend_type, '__call__'):
+    def detrend_func(d):
+      return detrend(d, type=detrend_type, axis=-1)
+  elif axis != -1:
+    # Wrap this function so that it receives a shape that it could
+    # reasonably expect to receive.
+    def detrend_func(d):
+      d = jnp.moveaxis(d, axis, -1)
+      d = detrend_type(d)
+      return jnp.moveaxis(d, -1, axis)
+  else:
+    detrend_func = detrend_type
+
+  if jnp.result_type(win, np.complex64) != outdtype:
+    win = win.astype(outdtype)
+
+  # Determine scale
+  if scaling == 'density':
+    scale = 1.0 / (fs * (win*win).sum())
+  elif scaling == 'spectrum':
+    scale = 1.0 / win.sum()**2
+  else:
+    raise ValueError(f'Unknown scaling: {scaling}')
+  if mode == 'stft':
+    scale = jnp.sqrt(scale)
+
+  # Determine onesided/ two-sided
+  if return_onesided:
+    if np.iscomplexobj(x):
+      sides = 'twosided'
+      warnings.warn('Input data is complex, switching to '
+                    'return_onesided=False')
+    else:
+      sides = 'onesided'
+      if y is not None:
+        if np.iscomplexobj(y):
+          sides = 'twosided'
+          warnings.warn('Input data is complex, switching to '
+                        'return_onesided=False')
+  else:
+    sides = 'twosided'
+
+  if sides == 'twosided':
+    freqs = jax.numpy.fft.fftfreq(nfft, 1/fs)
+  elif sides == 'onesided':
+    freqs = jax.numpy.fft.rfftfreq(nfft, 1/fs)
+
+  # Perform the windowed FFTs
+  result = _fft_helper(x, win, detrend_func, nperseg, noverlap, nfft, sides)
+
+  if y is not None:
+    # All the same operations on the y data
+    result_y = _fft_helper(y, win, detrend_func, nperseg, noverlap, nfft,
+                           sides)
+    result = jnp.conjugate(result) * result_y
+  elif mode == 'psd':
+    result = jnp.conjugate(result) * result
+
+  result *= scale
+
+  if sides == 'onesided' and mode == 'psd':
+    end = None if nfft % 2 else -1
+    result = result.at[..., 1:end].mul(2)
+
+  time = jnp.arange(nperseg/2, x.shape[-1] - nperseg/2 + 1,
+                    nperseg - noverlap)/float(fs)
```

Comment:
This casting of `fs` to `float` means that `fs` cannot be a traced quantity. We can probably remove that? I suspect this is left over from Python 2.X, where integer division was treated as floor division, so you had to make certain one od the inputs was floating point, but that's no longer necessary in Python 3.

Reply:
thanks for catching that. fixed.